UI: use dependency injection for factory hierarchy

klees opened the pull request:
This implements dependency injection for the factory hierarchy of the UI-framework. On the one hand this will speed up things a little, as factory aren't freshly initialized on every call. On the other hand we may now inject more dependencies at the correct locations in the hierarchy, e.g. data factory for inputs. There will be a follow up PR regarding these.

klees committed:
UI: use dependency injection for factory hierarchy

Amstutz reviewed (changes requested):
So you propose to wire everything up externally (e.g. in the Initialisation of ILIAS). Note that this makes the effort of adding new components higher, since one has to remember, to wire things up there as well. Please adapt the according .md (the section in which we explain step by step how new components are added). 

So we kind of inverse the pattern completely for instatiating the UI components, instead of going from the outside, working our way in, we go from the inside working our way out. This changes the pattern on how you will access this services from different context, right? E.g. in tests you will probably not use return `new \ILIAS\UI\Implementation\Factory(...);` anymore (otherwise you would have to perform the complete wiring up) but access components directly like: `return new \ILIAS\UI\Implementation\Component\Glyph\Factory();`
correct?

Does this not break many of our current unit tests (e.g. in dropzones). I did not check, but if so, please fix them.

klees commented:
You are indeed correct, tests brake. Shame on me for not running them =( Will fix.

We could move the wiring back into the namespace itself, but currently I cannot see clearly how this approach would work out in general. So I would currently not take this step. Your hint about the effort is correct as well and I will adjust the *.md accordingly.

What in fact is reversed is the way, the factory is constructed, not the way the component are constructed. I.e., once I have some `UI/Factory` I can use it to construct the components like we currently do. But it is in fact harder to construct some factory. This would also be alleviated by moving the wiring into the namespace. On the other hand, the usage of the factory was also a critique from @d3r1w during his investigations of unit tests. So I imo the correct way forward is to instantiate only the factory that is needed in the test, and not the complete hierarchy.

Testing in fact was also the very reason to introduce the DI. I wanted to have `ilLanguage` as dependency for `Validation::Factory`, which in turn is a dependency of `Field\Factory`. To have `ilLanguage` there, I would have needed to pull it out of thin air (some global). This would have been very bad for the tests. With DI the factorys tell up front which dependencies they need.

Amstutz commented:
> On the other hand, the usage of the factory was also a critique from @d3r1w during his investigations of unit tests. So I imo the correct way forward is to instantiate only the factory that is needed in the test, and not the complete hierarchy.

I agree.

> Testing in fact was also the very reason to introduce the DI. I wanted to have ilLanguage as dependency for Validation::Factory, which in turn is a dependency of Field\Factory. To have ilLanguage there, I would have needed to pull it out of thin air (some global). This would have been very bad for the tests. With DI the factorys tell up front which dependencies they need.

+1

I approve the concept. Thx, for your work. As far as I can see, there are no changes on the interface. The tests break, because of their unspecified way of instatiating the factories. From mpv you can merge this to trunk as soon as you:

- [x] Fixed the tests
- [x] Adapted the readme

klees committed:
UI: class Icon is not meant to be instantiated directly

klees committed:
UI: fixed DI in tests

klees committed:
UI: added wiring of factory in ilInitialisation to readme

klees commented:
@Amstutz I fixed the tests and added a [step 18](https://github.com/ILIAS-eLearning/ILIAS/pull/1054/files#diff-b21b79212c6b60a21608e061eb2b4203R402) to the readme.

@smeyer-ilias Any opinion from your side?

Amstutz commented:
Thanks for adding the point in the readme. Resolve the conflicts, make sure to add new factories in the init. Do you want to announce this at the JF or merge directly?

smeyer-ilias reviewed (approved):
LGTM
Your decision whether to merge directly or announce in JF.

klees commented:
IMHO JF is not necessary, nothing in here is a breaking change in a public interface.

klees commented:
@Amstutz Fixed merge conflicts and tests.

alex40724 commented on src/UI/Implementation/Factory.php:
```diff
@@ -4,41 +4,175 @@
+		C\Input\Factory $input_factory,
+		C\Table\Factory $table_factory,
+		C\MessageBox\Factory $messagebox_factory
+	) {
```
What about using a optional dependency data object that may contain these factories. If factories are present in the dependeny data object they are type checked and used, if not we fall back to call "new ...". This way the signature won't blow up and single changes to the default factories can be injected easily. In the future we may hopefully use https://wiki.php.net/rfc/typed_properties_v2 to do proper typisation of these objects.

klees replied:
I do not see a reason to hide the dependencies this object really has. "Signature blow up" might be a problem for some method, I cannot see why this would be a problem here. In the end one never needs to instantiate the object in productive code.

The long list of deps rather is a symptom of another problem we have here: the top level of the component tree gets kind of messy an would benefit from some additional structure. This is also the strategy we should use to solve similar issues (many deps) in other components: break the component in smaller parts with less dependencies.

alex40724 replied:
I think maintainability is something you want also in your test code. Now the signature is spread already in

tests/UI/Component/Breadcrumbs/BreadcrumbsTest.php
tests/UI/Component/Card/CardTest.php
tests/UI/Component/Deck/DeckTest.php
tests/UI/Component/Dropzone/File/DropzoneRendererTest.php
tests/UI/Component/Legacy/LegacyTest.php
tests/UI/Component/MessageBox/MessageBoxTest.php
tests/UI/Component/Modal/ModalBase.php
tests/UI/Component/ViewControl/PaginationTest.php
tests/UI/Component/ViewControl/SortationTest.php
tests/UI/Component/ViewControl/ViewControlTest.php

All these test classes need to be rewritten each time a new top level node is introduced. And most of the test do not really depend on the other components. The factory does. A MockFactory or some other helper class could hide all this.

klees replied:
The maintainability is indeed better if dependencies are made explicit, because devs can then see the deps and PHP can warn you that something is missing very early in the process. I see your point about the spread of the signature to the tests, though, and agree on hiding that in some helper.

Amstutz commented:
I see the point @alex40724 is making here. I am also somewhat concerned with the huge footprint we leave here. However, I think this would be a further refactoring. @alex40724 are you fine with merging this or would you like to discuss the issue further?

alex40724 commented:
@Amstutz: No, I made my point. If you are fine with it as a coordinater this is ready to be merged.

Amstutz commented:
@klees is this ready to be merged?

klees commented:
@Amstutz This is (again =) ) ready to be merged. Fixed tests and stuff. Please merge this soon so I won't have to do this again. Thx!
